File: CSE101_Data-Structures-and-Algorithms/pa1/Lex.java
        ////
        ////
        ////            Programming Assignment #1
        ////                Integer List ADT
        ////        Created by: Nicholas Scagliotti
        ////                nscaglio@ucsc.edu
        ////                   CMPS 101
        ////                   Fall '18
        ////
        ////
import java.io.*;
import java.util.Scanner;

public class Lex{
	public static void main(String [] args) throws IOException{
		int lineCount = 0;
		Scanner input = null;
		PrintWriter output = null;

		if(args.length < 2){											//Checking to see if there are at least two arguments
			System.out.println("Usage: FileIO infile outfile");
			System.exit(1);
		}

		input = new Scanner(new File(args[0]));

		while(input.hasNextLine()){									//Counting the total lines in the scanned file
         input.nextLine();
         lineCount++;
      	}
      	input.close();

		String[] token = new String[lineCount];
		Scanner get = new Scanner(new File(args[0]));
		output = new PrintWriter(new FileWriter(args[1]));

		for(int i = 0; i < token.length; i++){						//Creates string array and fills it with contents from the file.
			token[i] = get.nextLine();								//Second array should be the same size as total lines counted.
		}
		get.close();
		output.close();

		List L = new List();										//Instantiating a new list		
		for(int i = 1; i < token.length; i++){
			if(L.index() < 0){									//For the case of an empty list
        	   		L.append(i - 1);
       			}
         		L.moveFront();
         		int j = i;
         		while(j > 0 && token[i].compareTo(token[L.get()]) >= 0){   //Comparing the values, searching for where token[i] will come before token[L.get()]
         			L.moveNext();
         			j--;
         		}
         		if(L.index() >= 0){
         			L.insertBefore(i);										//Insterting it before the value we found to come after our desired element
         		}
         		else{														//At this point we've reached the end of the array and know the current value must be the largest.
         			L.append(i);											//Simply call the append function to place it in the desired location.
         		} 
     		}

     	output = new PrintWriter(new FileWriter(args[1]));					//Printing to outfile
     	L.moveFront();														//Set cursor to front and print in order
     	int i = 0;
     	while(i < lineCount){
     		output.println(token[L.get()]);
     		L.moveNext();
     		i++;
     	}
     	output.close();
	}
}

File: CSE101/pa2/Lex.c
        ////
        ////            Programming Assignment #2
        ////         Integer List ADT (Written in C)
        ////        Created by: Nicholas Scagliotti
        ////                nscaglio@ucsc.edu
        ////                   CMPS 101
        ////                   Fall '18
        ////
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"List.h"
#define MAX_LEN 200

int main(int argc, char *argv[]){
        int i, count = 0;
        FILE *infile, *outfile;
        char line[MAX_LEN];

        if(argc != 3){                          //Checking to see there is the correct amount of arguments
                printf("Usage: %s <input file> <output file> \n", argv[0]);
                exit(1);
        }

        infile = fopen(argv[1], "r");
        outfile = fopen(argv[2], "w");

        if(infile == NULL){                     //Case for input file failing to be read
                printf("Unable to open file %s for reading. \n", argv[1]);
                exit(1);
        }

        if(outfile == NULL){                    //Case for output file failing to write
                printf("Unable to open file %s for writing. \n", argv[2]);
                exit(1);
        }

        while(fgets(line, MAX_LEN, infile)){    //Counting the total lines in the scanned file
                count++;
        }rewind(infile);

        char array[count][MAX_LEN];             //Allocating an array for the strings
        i = 0;

        while(fgets(line, MAX_LEN, infile) != NULL){    //Copy into the newly created array
                strcpy(array[i], line);
                i++;
        }

        List L = newList();                     //Instantiating a new list

        for(i = 1; i < count; i++){         //Insertion sort to alphabetize the input
                char *temp = array[i];
                int j = i;
                if(index(L) < 0){
                    append(L, 0);
                }
                moveBack(L);
                while(j >= 0 && strcmp(temp, array[get(L)]) < 0){
                        movePrev(L);
                        j--;
                }
                if(index(L) > 0){
                        insertAfter(L, i);
                }
                else{
                    prepend(L, i);
                }
        }

        moveFront(L);
        i = 0;
        while(i < count){
                fprintf(outfile, "%s\n", array[get(L)]);
                moveNext(L);
                i++;
        }

        fclose(infile);                         //Closing both in and out files
        fclose(outfile);

        freeList(&L);                           //Freeing the list
        return 0;
}